Dungeon.py

```diff
@@ -1,80 +1,115 @@
-import os
+from __future__ import annotations
+from collections.abc import Iterator
+from typing import TYPE_CHECKING, Optional, Any
 
-from Hints import HintArea
-from Utils import data_path
+if TYPE_CHECKING:
+    from Hints import HintArea
+    from Item import Item
+    from Region import Region
+    from World import World
 
 
 class Dungeon:
-    def __init__(self, world, name, hint):
-        self.world = world
-        self.name = name
-        self.hint = hint
-        self.regions = []
-        self.boss_key = []
-        self.small_keys = []
-        self.dungeon_items = []
-        self.silver_rupees = []
-
-        for region in world.regions:
-            if region.dungeon == self.name:
+    def __init__(self, world: World, name: str, hint: HintArea, regions: Optional[list[Region]] = None) -> None:
+        self.world: World = world
+        self.name: str = name
+        self.hint: HintArea = hint
+        self.regions: list[Region] = regions if regions is not None else []
+        self.boss_key: list[Item] = []
+        self.small_keys: list[Item] = []
+        self.dungeon_items: list[Item] = []
+        self.silver_rupees: list[Item] = []
+
+        if regions is None:
+            for region in world.regions:
+                if region.dungeon_name != self.name:
+                    continue
                 region.dungeon = self
                 self.regions.append(region)
 
+    def copy(self, *, copy_dict: Optional[dict[int, Any]] = None) -> Dungeon:
```

Comment:
`*` just means that any parameter to the right of it has to be a named parameter and can't be used as a positional parameter.

Basically, the change is that these copy functions will now make copies of all the shit attached to them. The `copy_dict` is where references to those new copies are stored. It is using the python `id` of the old version of the object as the key so each thing is only copied once. It's optional because if it is not provided, one is created inside of the first copy method call before being passed down to all the resulting copy calls. The result should be that the copies created at the end have references to the copies of all the things they referenced before instead of doing some of the funky shit that we were doing to make it work.

Reply:
It's still a little bit over my head but I think I understand well enough